### PR TITLE
Fallback to use pathname from fdmappings entry for msync

### PR DESCRIPTION
**Pull request by vtikoo**
### Summary
- fdmappings are not cleaned up when fd is closed. This enables msync operations after file close.
- on `close(fd)`, the fd field in the corresponding file mapping entries is invalidated. Any later msync uses `fdmapping.pathname` to open the file. 

Signed-off-by: Vikas Tikoo <vitikoo@microsoft.com>

### CODE REVIEW

File: kernel/mmanutils.c [hunk 1]
```diff
@@ -960,15 +960,46 @@ int myst_msync(void* addr, size_t length, int flags)
...
 
         for (size_t i = index; i < index + n; i++)
         {
             myst_fdmapping_t* p = &v.fdmappings[i];
-
-            if (p->used == MYST_FDMAPPING_USED)
-                ECHECK(_sync_file(p->fd, p->offset, page, PAGE_SIZE));
+            /* sync only for writable mappings */
+            ECHECK(myst_mman_get_prot(
+                &_mman, page, PAGE_SIZE, &prot, &consistent));
```
Comment (mingweishih) [inline kernel/mmanutils.c:974]: `consistent` is not used. Do we need a check for it as well?

File: kernel/mmanutils.c [hunk 2]
```diff
@@ -960,15 +960,46 @@ int myst_msync(void* addr, size_t length, int flags)
...
+                if (p->fd != -1) // valid fd
+                {
+                    // file may have been truncated, ignore failure
+                    _sync_file(p->fd, p->offset, page, PAGE_SIZE);
+                }
+                // invalidated fd, fall back on pathname
+                else
+                {
+                    /* Get file handle the first time, and when the pathname
+                     * field changes*/
```
Comment (mingweishih) [inline kernel/mmanutils.c:986]:
```suggestion
                     * field changes */
```

File: kernel/mmanutils.c [hunk 3]
```diff
@@ -960,15 +960,46 @@ int myst_msync(void* addr, size_t length, int flags)
...
+                    if (!prev_pathname ||
+                        !strcmp(prev_pathname, p->pathname->data))
```
Comment (mingweishih) [inline kernel/mmanutils.c:988]: Does this mean that we only sync on mappings that have the same pathname as the first entry? Just to help me understand, can you explain a bit more on this?

File: kernel/mmanutils.c (hunk 1 above)
Comment (vtikoo) [reply kernel/mmanutils.c:974]: Protection is being checked for a single page here, so it should always be consistent.

Comment (mingweishih) [reply kernel/mmanutils.c:974]: Ah, I see. In this case, you can simply pass NULL instead of `&consistent`

File: kernel/mmanutils.c (hunk 3 above)
Comment (vtikoo) [reply kernel/mmanutils.c:989]: The sync happens always, this if block checks whether we need to get a new handle. We fetch a file handle once for the mapping interval which belongs to the same pathname. I have expanded the comment as well.

File: kernel/mmanutils.c (hunk 1 above)
Comment (vtikoo) [reply kernel/mmanutils.c:974]: `myst_mman_get_prot` does not allow a nullable `consistent` argument currently.

**Comment by vtikoo** [comment]
Overruled by #770

**Pull request closed without merging**